[Remote clusters] Add new security model

sabarasaba opened the pull request:
Fixes: https://github.com/elastic/kibana/issues/154312

## Summary
This PR takes care of updating the UX for the creation of remote clusters so that after creating a new configuration we show the users the correspondant docs on how to setup authentication for it. Auth can be configured in two ways: either server TLS or mutual TLS, based on what the user selects we need to explain to them how to configure it.


##### Notes
With this PR I also fixed the EUI deprecations from the List,Edit and Add pages.

##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial` and serverless search solution with `yarn start`
2. Navigate to the remote clusters UI and create a new cluster config
3. Verify that adding a new config has the two steps wizard
4. Verify that the config is created after the second step is compleated
5. Verify that if the user is on at least enterprise license the API Keys method is shown, but if the user is on basic license only certificates should be shown.
6. Verify that if an error happen when creating the config, the user is sent to the first step of the wizard


<details>
<summary>Screenshots empty states</summary>

![Screenshot 2023-07-19 at 08 42 57](https://github.com/elastic/kibana/assets/1191206/8ba59f3a-95de-48f1-a5b8-b5fa6df2b85a)
![Screenshot 2023-07-19 at 08 43 32](https://github.com/elastic/kibana/assets/1191206/2e79e0b7-14e6-49f8-ad76-49e55f381c82)
![Screenshot 2023-07-19 at 08 43 54](https://github.com/elastic/kibana/assets/1191206/6a9c9147-dd7c-48d4-a963-85357d4faec1)
<img width="1504" alt="Screenshot 2023-08-01 at 17 03 43" src="https://github.com/elastic/kibana/assets/1191206/80d05413-c0b0-4670-b781-f0384cc56bb1">


</details>

<details>
<summary>Screenshots </summary>

// Adding configuration flow
![Screenshot 2023-07-19 at 08 44 07](https://github.com/elastic/kibana/assets/1191206/a91a5f35-138b-45c8-9d5b-d7b53599bcd7)
// With local form errors
![Screenshot 2023-07-19 at 09 31 28](https://github.com/elastic/kibana/assets/1191206/db95236b-f641-416c-80b9-94e35d02bf4b)
// With enterprise license
<img width="1506" alt="Screenshot 2023-08-01 at 17 01 12" src="https://github.com/elastic/kibana/assets/1191206/7f4d054e-e24b-471a-bad8-4759ea537c18">
// Without enterprise license
<img width="1495" alt="Screenshot 2023-08-01 at 17 02 03" src="https://github.com/elastic/kibana/assets/1191206/3cc5fe03-4027-40c4-934d-8798071c94cc">
// With API error
![Screenshot 2023-07-20 at 09 04 15](https://github.com/elastic/kibana/assets/1191206/999d03b5-fd03-460f-8de5-b2c7ba3c6c46)

<img width="1497" alt="Screenshot 2023-08-01 at 17 01 37" src="https://github.com/elastic/kibana/assets/1191206/325be12d-1a18-48f2-943e-b8b4d45c9fd6">


![Screenshot 2023-07-19 at 08 44 31](https://github.com/elastic/kibana/assets/1191206/d6bb5d2c-a4a6-4f99-9829-78f41fff76b9)

// Remote clusters list
![Screenshot 2023-07-19 at 08 44 45](https://github.com/elastic/kibana/assets/1191206/8bc2f4dd-ffeb-4570-9b86-2f597c4c90eb)

// Edit remote cluster config
![Screenshot 2023-07-19 at 08 44 51](https://github.com/elastic/kibana/assets/1191206/73bb5567-349d-46d6-b680-785cae64aeab)

</details>

alaudazzi commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/remote_cluster_setup_trust.tsx:
```diff
@@ -0,0 +1,181 @@
+        <p>
+          <FormattedMessage
+            id="xpack.remoteClusters.clusterWizard.trustStep.title"
+            defaultMessage="To establish trust between this deployment and your newly created remote cluster, {br} choose one of the two options below."
```
```suggestion
            defaultMessage="To establish trust between this deployment and your newly created remote cluster, {br} choose one of the following options."
```
Replace "below" with a more accessible term. Check [How to create accessible web content](https://docs.google.com/document/d/1IGuMf16-cwxx8BX1Z9PWsFWsTPxicwTkX_r7UOhjd4E/edit).

alaudazzi commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/confirm_modal.tsx:
```diff
@@ -0,0 +1,104 @@
+          <p>
+            <FormattedMessage
+              id="xpack.remoteClusters.clusterWizard.trustStep.body"
+              defaultMessage="Have you setup an authentication mechanism to securely connect your remote cluster?"
```
```suggestion
              defaultMessage="Have you setup trust to securely connect your remote cluster?"
```

alaudazzi reviewed (commented):
Added a few editing suggestions.

alaudazzi commented:
One more general comment:

<img width="838" alt="image" src="https://github.com/elastic/kibana/assets/46651782/5b1475eb-e908-404a-92f9-53c2d0d0cf12">

The three blocks marked in red basically repeat the same thing. You can probably consider to:
1.  Rewrite the introductory text with "**To establish trust between this deployment and your remote cluster, choose one of the following options.**"
2. Remove the text in the boxes and keep just **API Keys** and **Certificates**.

alaudazzi commented:
@sabarasaba I also left some comments in [figma](https://www.figma.com/file/rfmtg1EhRTMP21Alf0QbeE/Remote-clusters---security-2.0?type=design&node-id=328-218142&mode=design&t=EyvrJlESSiWlbi0O-0#487426153).

sabarasaba commented:
Thanks for the review @alaudazzi! I've addressed most of your feedback but here are two notes:

> Remove the text in the boxes and keep just API Keys and Certificates.

I think we should tell at least something to the user about the different authentication methods, otherwise they are left with having to go into the docs for each option and read more for each one of them before deciding. I think we could do better and give them a small summary of what each of the options are so that they can make a decition earlier on in the process. Wdy?

> Consider to rephrase the following text:
> 
> > Be aware that the API key authentication won't work with CCR.
> 
> with
> 
> > The API key authentication will not work with CCR.

As discussed in the remote cluster sync, this text is no longer needed hence why is not included in the PR

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/remote_cluster_setup_trust.tsx:
```diff
@@ -0,0 +1,182 @@
+  apiKeyCloud: remoteClustersUrl,
+  cert: remoteClustersUrl,
+  certCloud: remoteClustersUrl,
+};
```
These docs links are just hardcoded for now to be all the same, but once the docs team finish off the doc pages we can come back to to patch this,

sabarasaba commented:
@elasticmachine merge upstream

alaudazzi commented:
> I think we should tell at least something to the user about the different authentication methods, otherwise they are left with having to go into the docs for each option and read more for each one of them before deciding. I think we could do better and give them a small summary of what each of the options are so that they can make a decition earlier on in the process. Wdy?

Makes sense. However, the text as is now seems to add no value on top of what is already stated above the two boxes. The text in the boxes should call out the main difference(s) between the two options.

@bytebilly @ywangd maybe you can help here?

ElenaStoeva commented:
Nice work @sabarasaba, the new form wizard looks great! 🎉 

Tested locally in classic Kibana and everything works as expected (the "How to test" instructions say to test this in serverless but I suppose this is a mistake since Remote clusters plugin is disabled in serverless).

Code changes overall look good to me but I'll take a closer look next week.

ywangd commented:
> The text in the boxes should call out the main difference(s) between the two options.

I agree. Though it might be challenging to capture it nicely and briefly. On the technical side, the main difference between the two models to me is whether the two clusters are considered as controled by the same admin. The "API key model" does _not_ assume so while the "certificate model" does (which means a superuser is on one cluster is also a superuser on the other cluster). Not sure whether this is the right level of concept for end-users. I'll let @bytebilly to comment as well.

ElenaStoeva reviewed (commented):
Hi @sabarasaba, I reviewed the code changes in more detail and I only have two small comments.

alisonelizabeth reviewed (commented):
Thanks for working on this @sabarasaba! I pulled down to take a quick look and also grab a screenshot for a team update email. 

Can you remind me again how to successfully set up a remote cluster? 😅 I'm following the [instructions here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/cross_cluster_replication/README.md#quick-steps-for-testing-cross-cluster-replication), but I keep getting the "Not connected" status.

I'll defer to design/docs, but I also have some minor feedback around the UX:

- Should the button copy be the same? I also think on the list view, we use "Add remote cluster".

<img width="965" alt="Screenshot 2023-07-24 at 2 13 40 PM" src="https://github.com/elastic/kibana/assets/5226211/465f0828-84d0-4568-a972-f5c3a6fbc97c">

- Two thoughts on the cards:
1. I think we need to change the text when only one option is shown
2. In general, I think the copy needs to be tweaked a bit here. It says "Choose one" but then the card is not clickable and there's only a "Read more" link. Maybe a quick fix would be to change the "Read more" link copy to something more actionable?

<img width="627" alt="Screenshot 2023-07-24 at 2 19 22 PM" src="https://github.com/elastic/kibana/assets/5226211/3e537aa0-6843-4f18-9481-26f8a51cb626">

- This is probably an old bug, but I _think_ the background should be red for an error state.
<img width="503" alt="Screenshot 2023-07-24 at 2 45 46 PM" src="https://github.com/elastic/kibana/assets/5226211/fe011367-c636-4fa2-a3b9-e8edbc1b348e">

sabarasaba commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/confirm_modal.tsx:
```diff
@@ -0,0 +1,104 @@
+          <p>
+            <FormattedMessage
+              id="xpack.remoteClusters.clusterWizard.trustStep.body"
+              defaultMessage="Have you setup trust to connect to your remote cluster?"
```
I think you might be right, but I'll leave that up to @alaudazzi

alaudazzi replied:
Yes, **set up** is the verb **setup** is the noun.

sabarasaba commented on x-pack/plugins/remote_clusters/public/application/sections/remote_cluster_add/wizard_form.tsx:
```diff
@@ -0,0 +1,102 @@
+          defaultMessage: 'Add connection information',
+        }),
+        status: (currentStep === CONFIGURE_CONNECTION ? 'current' : 'complete') as EuiStepStatus,
+        onClick: () => {},
```
I've reached out to @kyrspl to confirm, but was my understanding that these were not going to be clickeable. But the `onClick` handler is a required prop for `EuiStepHorizontal` which is why I left it as an empty function here.

alaudazzi replied:
@alisonelizabeth @sabarasaba 
1. The text "choose one of the following options" was suggested when the two options  -- API and Certificates appeared on the screen. 
2. The text on the **Certificates** tab sounds weird. If I recall correctly from the last conversation we had with Ignacio, we don't want to convey the idea of a "less secure" method. @sabarasaba please confirm?

sabarasaba commented on x-pack/test/functional/page_objects/remote_clusters_page.ts:
```diff
@@ -29,6 +29,12 @@ export function RemoteClustersPageProvider({ getService }: FtrProviderContext) {
       await testSubjects.setValue('remoteClusterFormNameInput', name);
       await comboBox.setCustom('comboBoxInput', seedNode);
+
+      // Complete trust setup
```
Good catch! I've moved the calls around to accomodate that, the skipped tests are part of another issue https://github.com/elastic/kibana/issues/153599 but I think is a bit out of scope for this PR so I'll leave it skipped and we can pick it up after.

ElenaStoeva replied:
Yes, the test was skipped because of an accessibility issue, but the `createNewRemoteCluster()` function which is part of the test setup was working correctly before these changes - it fails now because it cannot navigate the "Yes, I have setup trust" checkbox.

I was thinking that we could leave the test skipped but make sure that the `createNewRemoteCluster()` function executes successfully as before. Wdyt?

sabarasaba replied:
Yes! I've patched it up but haven't pushed changes yet as I'm addressing other feedback I got for this pr!

ElenaStoeva replied:
Got it, thanks!

alaudazzi commented:
@florent-leborgne might probably chime in and provide his feedback?

bytebilly commented:
> The text in the boxes should call out the main difference(s) between the two options.

I agree that we can help users understand the right model for them with some inline text.
As @ywangd said, the new model is suitable when clusters don't share the same security domain. However, it can also work well when they share the same security domain. The only reason why we keep exposing the old model is that the new one is beta, and I expect that option to go away as soon as the new one will be GA.

This means that we can focus in finding the right way to expose both, but it will last (hopefully) just one or two releases. After that, we'll always have a single choice (based on the available license), and listing differences doesn't add a lot of value anymore.

I suggest the following text:

> **API keys**
> Fine-grained access control on remote indices. Use this option with an API key provided by the remote cluster administrator.

> **Certificates**
> Full access to the remote cluster. Use this option if you have TLS certificates from the remote cluster.

In GA, we could potentially list instructions for the "right" security model instead of showing the box.
What do you think?

sabarasaba commented:
> I suggest the following text:
> 
> > **API keys**
> > Fine-grained access control on remote indices. Use this option with an API key provided by the remote cluster administrator.
> 
> > **Certificates**
> > Full access to the remote cluster. Use this option if you have TLS certificates from the remote cluster.
> 
> In GA, we could potentially list instructions for the "right" security model instead of showing the box. What do you think?

This makes sense for me, I can make the changes if @florent-leborgne @alaudazzi approve it too.

florent-leborgne reviewed (commented):
A few comments and suggestions on the wording. Feel free to ignore what's not relevant, I'mp probably missing a bit of context or discussion history on a few things :)

sabarasaba commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/remote_cluster_setup_trust.tsx:
```diff
@@ -0,0 +1,182 @@
+        <p>
+          <FormattedMessage
+            id="xpack.remoteClusters.clusterWizard.trustStep.title"
+            defaultMessage="To establish trust between this deployment and your remote cluster, {br} choose one of the following options."
```
Thats a great idea, love the suggestion!

sabarasaba commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/remote_cluster_setup_trust.tsx:
```diff
@@ -0,0 +1,182 @@
+              >
+                <FormattedMessage
+                  id="xpack.remoteClusters.clusterWizard.trustStep.docs"
+                  defaultMessage="Read more"
```
Ill defer this to @kyrspl since I'm not too sure whats best

sabarasaba commented on x-pack/plugins/remote_clusters/public/application/sections/remote_cluster_add/remote_cluster_add.js:
```diff
@@ -56,29 +58,31 @@ export class RemoteClusterAdd extends PureComponent {
+            description={
+              <FormattedMessage
+                id="xpack.remoteClusters.remoteClustersDescription"
+                defaultMessage="Add a remote cluster that connects to seed nodes or to a single proxy address."
```
Love the suggestion, a lot clearer now!

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
Thanks y'all for your feedback I've addressed most of your comments with the latest commit https://github.com/elastic/kibana/pull/161836/commits/abfe04696387954fae85ce2a47d2ce7224989077. @alisonelizabeth 
 I've included instructions on the readme on how to successfully connect to a "local" remote cluster.

kyrspl commented on x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_setup_trust/remote_cluster_setup_trust.tsx:
```diff
@@ -0,0 +1,182 @@
+            >
+              <FormattedMessage
+                id="xpack.remoteClusters.clusterWizard.trustStep.docs"
+                defaultMessage="Read more"
```
It is not mandatory. It's more like a suggested approach. If the user doesn't follow it, the cluster will still be created.

@florent-leborgne if you feel this is a more actionable CTA then by all means, let's use the proposed text.

florent-leborgne replied:
But it seems that we really want them to follow the docs and set it up at that moment, since we after have the modal where we're asking them to confirm they did set it up, so I think it makes more sense to have the doc links as an actionable CTA.

sabarasaba replied:
I've made the copy changes and transformed the links into ctas

![Screenshot 2023-07-26 at 07 31 24](https://github.com/elastic/kibana/assets/1191206/ab9a95a1-8de2-424a-94ab-dbe2fc250c64)

kyrspl commented:
> I'll defer to design/docs, but I also have some minor feedback around the UX:
> 
> * Should the button copy be the same? I also think on the list view, we use "Add remote cluster".
> 
> <img alt="Screenshot 2023-07-24 at 2 13 40 PM" width="965" src="https://user-images.githubusercontent.com/5226211/255686677-465f0828-84d0-4568-a972-f5c3a6fbc97c.png">

Here I suggest to use the same CTA text as we use in the previous page where we see all remote clusters. 

> * Two thoughts on the cards:
> 
> 1. I think we need to change the text when only one option is shown
> 2. In general, I think the copy needs to be tweaked a bit here. It says "Choose one" but then the card is not clickable and there's only a "Read more" link. Maybe a quick fix would be to change the "Read more" link copy to something more actionable?
> 
> <img alt="Screenshot 2023-07-24 at 2 19 22 PM" width="627" src="https://user-images.githubusercontent.com/5226211/255687245-3e537aa0-6843-4f18-9481-26f8a51cb626.png">
> * This is probably an old bug, but I _think_ the background should be red for an error state.
> 
> <img alt="Screenshot 2023-07-24 at 2 45 46 PM" width="503" src="https://user-images.githubusercontent.com/5226211/255691745-fe011367-c636-4fa2-a3b9-e8edbc1b348e.png">

Good catch @alisonelizabeth. 

The [empty prompt](https://elastic.github.io/eui/#/display/empty-prompt) does have an error state: `color`  can be set to `danger`.

sabarasaba commented:
Thanks for the reviews @florent-leborgne and @kyrspl! I've addressed your latest feedback with https://github.com/elastic/kibana/pull/161836/commits/8434948d730ff89f4ab3b24405a823656e508f60. 

Sadly it looks like this project won't go out for 8.10 as we expected and might delay to a further version and might involve more UI changes given requirement changes in ES side. I'll keep this PR open for the time being to see how things develop next week.

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

florent-leborgne reviewed (approved):
Content LGTM 🚀, besides the modal title that we agreed to revert @sabarasaba to avoid making it too repetitive.
Thanks for the changes!

alisonelizabeth commented on x-pack/plugins/remote_clusters/README.md:
```diff
@@ -1,5 +1,13 @@
 # Remote Clusters
 
+## Setting up a remote cluster
```
Thanks for adding this! I'm still not able to get a remote cluster with a successful connection 🤔. I noticed a few errors in the ES logs related to the license, `...operation due to expired license.`, so might be related to that.

sabarasaba replied:
Ohhh thats odd, I was able to connect it consistently using this method 🤔

alisonelizabeth reviewed (approved):
Thanks for working on this @sabarasaba! Latest LGTM. Tested locally. I think the copy is much cleaner now too 👍

alisonelizabeth commented:
@sabarasaba can we also check off the `remote_clusters` plugin work from https://github.com/elastic/kibana/issues/161413 as part of this PR?

sabarasaba commented:
> @sabarasaba can we also check off the `remote_clusters` plugin work from #161413 as part of this PR?

Yes! This pr addresses the two instances where we were using deprecations in the RC plugin.

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
buildkite test this

sabarasaba commented:
@elasticmachine merge upstream

florent-leborgne commented:
@sabarasaba I've seen [on slack](https://elastic.slack.com/archives/C04NYQNA6LX/p1691483187183669) that the API key mechanism is only possible if the remote cluster is also at least on 8.10

Should we mention it as well on the API key card?  
<img width="210" alt="image" src="https://github.com/elastic/kibana/assets/10208282/2b3bd683-46f3-4395-9a19-8cbba1477579">

sabarasaba commented:
> @sabarasaba I've seen on slack that the API key mechanism is only possible if the remote cluster is also at least on 8.10
> 
> Should we mention it as well on the API key card? <img alt="image" width="210" src="https://user-images.githubusercontent.com/10208282/259108046-2b3bd683-46f3-4395-9a19-8cbba1477579.png">

Just saw that! Do you have any suggestions on how we could update the copy to reflect that? 🤔

kyrspl commented:
> Just saw that! Do you have any suggestions on how we could update the copy to reflect that? 🤔

We could use the `muted` option underneath the CTA to specify which version they work on. It's not the most visible option of course.

sabarasaba commented:
@elasticmachine merge upstream

abdonpijpelink reviewed (approved):
Doc links look good, but result in a 404 until https://github.com/elastic/elasticsearch/pull/98330 gets merged. I'll set up temporary redirects for them so this PR is unblocked.
